// workbox-google-analytics/src/utils/dist/constants.js
"use strict";
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
exports.__esModule = true;
exports.COLLECT_PATHS_REGEX = exports.COLLECT_DEFAULT_PATH = exports.GTM_JS_PATH = exports.GTAG_JS_PATH = exports.ANALYTICS_JS_PATH = exports.GTM_HOST = exports.GOOGLE_ANALYTICS_HOST = exports.MAX_RETENTION_TIME = exports.QUEUE_NAME = void 0;
require("../_version.js");
exports.QUEUE_NAME = 'workbox-google-analytics';
exports.MAX_RETENTION_TIME = 60 * 48; // Two days in minutes
exports.GOOGLE_ANALYTICS_HOST = 'www.google-analytics.com';
exports.GTM_HOST = 'www.googletagmanager.com';
exports.ANALYTICS_JS_PATH = '/analytics.js';
exports.GTAG_JS_PATH = '/gtag/js';
exports.GTM_JS_PATH = '/gtm.js';
exports.COLLECT_DEFAULT_PATH = '/collect';
// This RegExp matches all known Measurement Protocol single-hit collect
// endpoints. Most of the time the default path (/collect) is used, but
// occasionally an experimental endpoint is used when testing new features,
// (e.g. /r/collect or /j/collect)
exports.COLLECT_PATHS_REGEX = /^\/(\w+\/)?collect/;
