// @babel/core/lib/config/dist/resolve-targets.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resolveBrowserslistConfigFile = resolveBrowserslistConfigFile;
exports.resolveTargets = resolveTargets;

function _path() {
  var data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

function _helperCompilationTargets() {
  var data = require("@babel/helper-compilation-targets");

  _helperCompilationTargets = function _helperCompilationTargets() {
    return data;
  };

  return data;
}

({});

function resolveBrowserslistConfigFile(browserslistConfigFile, configFileDir) {
  return _path().resolve(configFileDir, browserslistConfigFile);
}

function resolveTargets(options, root) {
  var targets = options.targets;

  if (typeof targets === "string" || Array.isArray(targets)) {
    targets = {
      browsers: targets
    };
  }

  if (targets && targets.esmodules) {
    targets = Object.assign({}, targets, {
      esmodules: "intersect"
    });
  }

  var browserslistConfigFile = options.browserslistConfigFile;
  var configFile;
  var ignoreBrowserslistConfig = false;

  if (typeof browserslistConfigFile === "string") {
    configFile = browserslistConfigFile;
  } else {
    ignoreBrowserslistConfig = browserslistConfigFile === false;
  }

  return (0, _helperCompilationTargets()["default"])(targets, {
    ignoreBrowserslistConfig: ignoreBrowserslistConfig,
    configFile: configFile,
    configPath: root,
    browserslistEnv: options.browserslistEnv
  });
}
