// rollup-pluginutils/src/dist/extractAssignedNames.js
"use strict";
exports.__esModule = true;
exports["default"] = void 0;
var extractors = {
    ArrayPattern: function (names, param) {
        for (var _i = 0, _a = param.elements; _i < _a.length; _i++) {
            var element = _a[_i];
            if (element)
                extractors[element.type](names, element);
        }
    },
    AssignmentPattern: function (names, param) {
        extractors[param.left.type](names, param.left);
    },
    Identifier: function (names, param) {
        names.push(param.name);
    },
    MemberExpression: function () { },
    ObjectPattern: function (names, param) {
        for (var _i = 0, _a = param.properties; _i < _a.length; _i++) {
            var prop = _a[_i];
            if (prop.type === 'RestElement') {
                extractors.RestElement(names, prop);
            }
            else {
                extractors[prop.value.type](names, prop.value);
            }
        }
    },
    RestElement: function (names, param) {
        extractors[param.argument.type](names, param.argument);
    }
};
var extractAssignedNames = function extractAssignedNames(param) {
    var names = [];
    extractors[param.type](names, param);
    return names;
};
exports["default"] = extractAssignedNames;
