// workbox-routing/src/dist/index.js
"use strict";
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
exports.__esModule = true;
exports.setDefaultHandler = exports.setCatchHandler = exports.Router = exports.Route = exports.registerRoute = exports.RegExpRoute = exports.NavigationRoute = void 0;
var NavigationRoute_js_1 = require("./NavigationRoute.js");
exports.NavigationRoute = NavigationRoute_js_1.NavigationRoute;
var RegExpRoute_js_1 = require("./RegExpRoute.js");
exports.RegExpRoute = RegExpRoute_js_1.RegExpRoute;
var registerRoute_js_1 = require("./registerRoute.js");
exports.registerRoute = registerRoute_js_1.registerRoute;
var Route_js_1 = require("./Route.js");
exports.Route = Route_js_1.Route;
var Router_js_1 = require("./Router.js");
exports.Router = Router_js_1.Router;
var setCatchHandler_js_1 = require("./setCatchHandler.js");
exports.setCatchHandler = setCatchHandler_js_1.setCatchHandler;
var setDefaultHandler_js_1 = require("./setDefaultHandler.js");
exports.setDefaultHandler = setDefaultHandler_js_1.setDefaultHandler;
require("./_version.js");
