// workbox-core/src/dist/cacheNames.js
"use strict";
/*
  Copyright 2019 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
exports.__esModule = true;
exports.cacheNames = void 0;
var cacheNames_js_1 = require("./_private/cacheNames.js");
require("./_version.js");
/**
 * Get the current cache names and prefix/suffix used by Workbox.
 *
 * `cacheNames.precache` is used for precached assets,
 * `cacheNames.googleAnalytics` is used by `workbox-google-analytics` to
 * store `analytics.js`, and `cacheNames.runtime` is used for everything else.
 *
 * `cacheNames.prefix` can be used to retrieve just the current prefix value.
 * `cacheNames.suffix` can be used to retrieve just the current suffix value.
 *
 * @return {Object} An object with `precache`, `runtime`, `prefix`, and
 *     `googleAnalytics` properties.
 *
 * @memberof module:workbox-core
 */
var cacheNames = {
    get googleAnalytics() {
        return cacheNames_js_1.cacheNames.getGoogleAnalyticsName();
    },
    get precache() {
        return cacheNames_js_1.cacheNames.getPrecacheName();
    },
    get prefix() {
        return cacheNames_js_1.cacheNames.getPrefix();
    },
    get runtime() {
        return cacheNames_js_1.cacheNames.getRuntimeName();
    },
    get suffix() {
        return cacheNames_js_1.cacheNames.getSuffix();
    }
};
exports.cacheNames = cacheNames;
