// workbox-core/src/dist/index.js
"use strict";
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
exports.__esModule = true;
exports.skipWaiting = exports.setCacheNameDetails = exports.registerQuotaErrorCallback = exports.copyResponse = exports.clientsClaim = exports.cacheNames = exports._private = void 0;
var registerQuotaErrorCallback_js_1 = require("./registerQuotaErrorCallback.js");
exports.registerQuotaErrorCallback = registerQuotaErrorCallback_js_1.registerQuotaErrorCallback;
var _private = require("./_private.js");
exports._private = _private;
var cacheNames_js_1 = require("./cacheNames.js");
exports.cacheNames = cacheNames_js_1.cacheNames;
var copyResponse_js_1 = require("./copyResponse.js");
exports.copyResponse = copyResponse_js_1.copyResponse;
var clientsClaim_js_1 = require("./clientsClaim.js");
exports.clientsClaim = clientsClaim_js_1.clientsClaim;
var setCacheNameDetails_js_1 = require("./setCacheNameDetails.js");
exports.setCacheNameDetails = setCacheNameDetails_js_1.setCacheNameDetails;
var skipWaiting_js_1 = require("./skipWaiting.js");
exports.skipWaiting = skipWaiting_js_1.skipWaiting;
require("./_version.js");
__exportStar(require("./types.js"), exports);
