// @babel/helper-define-polyfill-provider/lib/dist/utils.dev.js
"use strict";

function _templateObject3() {
  var data = _taggedTemplateLiteral(["var ", " = require(", ")"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n                  var ", " = require(", ").", "\n                "]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["require(", ")"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

exports.__esModule = true;
exports.intersection = intersection;
exports.has = has;
exports.resolveKey = resolveKey;
exports.resolveSource = resolveSource;
exports.getImportSource = getImportSource;
exports.getRequireSource = getRequireSource;
exports.createUtilsGetter = createUtilsGetter;

var babel = _interopRequireWildcard(require("@babel/core"));

function _getRequireWildcardCache() {
  if (typeof WeakMap !== "function") return null;
  var cache = new WeakMap();

  _getRequireWildcardCache = function _getRequireWildcardCache() {
    return cache;
  };

  return cache;
}

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj;
  }

  if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {
    return {
      "default": obj
    };
  }

  var cache = _getRequireWildcardCache();

  if (cache && cache.has(obj)) {
    return cache.get(obj);
  }

  var newObj = {};
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;

  for (var key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;

      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }

  newObj["default"] = obj;

  if (cache) {
    cache.set(obj, newObj);
  }

  return newObj;
}

var _ref = babel["default"] || babel,
    t = _ref.types,
    template = _ref.template;

function intersection(a, b) {
  var result = new Set();
  a.forEach(function (v) {
    return b.has(v) && result.add(v);
  });
  return result;
}

function has(object, key) {
  return Object.prototype.hasOwnProperty.call(object, key);
}

function getType(target) {
  return Object.prototype.toString.call(target).slice(8, -1);
}

function resolveId(path) {
  if (path.isIdentifier() && !path.scope.hasBinding(path.node.name,
  /* noGlobals */
  true)) {
    return path.node.name;
  }

  var _path$evaluate = path.evaluate(),
      deopt = _path$evaluate.deopt;

  if (deopt && deopt.isIdentifier()) {
    return deopt.node.name;
  }
}

function resolveKey(path) {
  var computed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  var node = path.node,
      parent = path.parent,
      scope = path.scope;
  if (path.isStringLiteral()) return node.value;
  var name = node.name;
  var isIdentifier = path.isIdentifier();
  if (isIdentifier && !(computed || parent.computed)) return name;

  if (computed && path.isMemberExpression() && path.get("object").isIdentifier({
    name: "Symbol"
  }) && !scope.hasBinding("Symbol",
  /* noGlobals */
  true)) {
    var sym = resolveKey(path.get("property"), path.node.computed);
    if (sym) return "Symbol." + sym;
  }

  if (!isIdentifier || scope.hasBinding(name,
  /* noGlobals */
  true)) {
    var _path$evaluate2 = path.evaluate(),
        value = _path$evaluate2.value;

    if (typeof value === "string") return value;
  }
}

function resolveSource(obj) {
  if (obj.isMemberExpression() && obj.get("property").isIdentifier({
    name: "prototype"
  })) {
    var _id = resolveId(obj.get("object"));

    if (_id) {
      return {
        id: _id,
        placement: "prototype"
      };
    }

    return {
      id: null,
      placement: null
    };
  }

  var id = resolveId(obj);

  if (id) {
    return {
      id: id,
      placement: "static"
    };
  }

  var _obj$evaluate = obj.evaluate(),
      value = _obj$evaluate.value;

  if (value !== undefined) {
    return {
      id: getType(value),
      placement: "prototype"
    };
  } else if (obj.isRegExpLiteral()) {
    return {
      id: "RegExp",
      placement: "prototype"
    };
  } else if (obj.isFunction()) {
    return {
      id: "Function",
      placement: "prototype"
    };
  }

  return {
    id: null,
    placement: null
  };
}

function getImportSource(_ref2) {
  var node = _ref2.node;
  if (node.specifiers.length === 0) return node.source.value;
}

function getRequireSource(_ref3) {
  var node = _ref3.node;
  if (!t.isExpressionStatement(node)) return;
  var expression = node.expression;
  var isRequire = t.isCallExpression(expression) && t.isIdentifier(expression.callee) && expression.callee.name === "require" && expression.arguments.length === 1 && t.isStringLiteral(expression.arguments[0]);
  if (isRequire) return expression.arguments[0].value;
}

function hoist(node) {
  node._blockHoist = 3;
  return node;
}

function createUtilsGetter(cache) {
  return function (path) {
    var prog = path.findParent(function (p) {
      return p.isProgram();
    });
    return {
      injectGlobalImport: function injectGlobalImport(url) {
        cache.storeAnonymous(prog, url, function (isScript, source) {
          return isScript ? template.statement.ast(_templateObject(), source) : t.importDeclaration([], source);
        });
      },
      injectNamedImport: function injectNamedImport(url, name) {
        var hint = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : name;
        return cache.storeNamed(prog, url, name, function (isScript, source, name) {
          var id = prog.scope.generateUidIdentifier(hint);
          return {
            node: isScript ? hoist(template.statement.ast(_templateObject2(), id, source, name)) : t.importDeclaration([t.importSpecifier(id, name)], source),
            name: id.name
          };
        });
      },
      injectDefaultImport: function injectDefaultImport(url) {
        var hint = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : url;
        return cache.storeNamed(prog, url, "default", function (isScript, source) {
          var id = prog.scope.generateUidIdentifier(hint);
          return {
            node: isScript ? hoist(template.statement.ast(_templateObject3(), id, source)) : t.importDeclaration([t.importDefaultSpecifier(id)], source),
            name: id.name
          };
        });
      }
    };
  };
}
