// optimize-css-assets-webpack-plugin/test/util/dist/helpers.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readFileOrEmpty = readFileOrEmpty;
exports.checkForWebpackErrors = checkForWebpackErrors;
exports.defaultConfig = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _extractTextWebpackPlugin = _interopRequireDefault(require("extract-text-webpack-plugin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function readFileOrEmpty(path) {
  try {
    return _fs["default"].readFileSync(path, "utf-8");
  } catch (e) {
    return "";
  }
}

var defaultConfig = {
  entry: "./index",
  module: {
    rules: [{
      test: /\.css$/,
      use: _extractTextWebpackPlugin["default"].extract({
        fallback: {
          loader: "style-loader"
        },
        use: {
          loader: "css-loader"
        }
      })
    }]
  },
  plugins: [],
  context: __dirname,
  output: {
    filename: "destination.js",
    path: _path["default"].resolve(__dirname, "../", "js", "default-exports")
  }
};
exports.defaultConfig = defaultConfig;

function checkForWebpackErrors(_ref) {
  var err = _ref.err,
      stats = _ref.stats,
      done = _ref.done;
  if (err) return done(err);
  if (stats.hasErrors()) return done(new Error(stats.toString()));
}
