// react-dev-utils/dist/typescriptFormatter.dev.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
'use strict';

var os = require('os');

var codeFrame = require('@babel/code-frame').codeFrameColumns;

var chalk = require('chalk');

var fs = require('fs');

var issueOrigins = {
  typescript: 'TypeScript',
  internal: 'fork-ts-checker-webpack-plugin'
};

function formatter(issue) {
  var origin = issue.origin,
      severity = issue.severity,
      file = issue.file,
      line = issue.line,
      message = issue.message,
      code = issue.code,
      character = issue.character;
  var colors = new chalk.constructor();
  var messageColor = severity === 'warning' ? colors.yellow : colors.red;
  var fileAndNumberColor = colors.bold.cyan;
  var source = file && fs.existsSync(file) && fs.readFileSync(file, 'utf-8');
  var frame = source ? codeFrame(source, {
    start: {
      line: line,
      column: character
    }
  }).split('\n').map(function (str) {
    return '  ' + str;
  }).join(os.EOL) : '';
  return [messageColor.bold("".concat(issueOrigins[origin], " ").concat(severity.toLowerCase(), " in ")) + fileAndNumberColor("".concat(file, "(").concat(line, ",").concat(character, ")")) + messageColor(':'), message + '  ' + messageColor.underline("TS".concat(code)), '', frame].join(os.EOL);
}

module.exports = formatter;
