// workbox-precaching/src/dist/index.js
"use strict";
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
exports.__esModule = true;
exports.PrecacheController = exports.precacheAndRoute = exports.precache = exports.matchPrecache = exports.getCacheKeyForURL = exports.createHandlerBoundToURL = exports.createHandler = exports.cleanupOutdatedCaches = exports.addRoute = exports.addPlugins = void 0;
var addPlugins_js_1 = require("./addPlugins.js");
exports.addPlugins = addPlugins_js_1.addPlugins;
var addRoute_js_1 = require("./addRoute.js");
exports.addRoute = addRoute_js_1.addRoute;
var cleanupOutdatedCaches_js_1 = require("./cleanupOutdatedCaches.js");
exports.cleanupOutdatedCaches = cleanupOutdatedCaches_js_1.cleanupOutdatedCaches;
var createHandler_js_1 = require("./createHandler.js");
exports.createHandler = createHandler_js_1.createHandler;
var createHandlerBoundToURL_js_1 = require("./createHandlerBoundToURL.js");
exports.createHandlerBoundToURL = createHandlerBoundToURL_js_1.createHandlerBoundToURL;
var getCacheKeyForURL_js_1 = require("./getCacheKeyForURL.js");
exports.getCacheKeyForURL = getCacheKeyForURL_js_1.getCacheKeyForURL;
var matchPrecache_js_1 = require("./matchPrecache.js");
exports.matchPrecache = matchPrecache_js_1.matchPrecache;
var precache_js_1 = require("./precache.js");
exports.precache = precache_js_1.precache;
var precacheAndRoute_js_1 = require("./precacheAndRoute.js");
exports.precacheAndRoute = precacheAndRoute_js_1.precacheAndRoute;
var PrecacheController_js_1 = require("./PrecacheController.js");
exports.PrecacheController = PrecacheController_js_1.PrecacheController;
require("./_version.js");
