// mini-css-extract-plugin/dist/dist/CssModule.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _webpack = _interopRequireDefault(require("webpack"));

var _utils = require("./utils");

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

var TYPES = new Set([_utils.MODULE_TYPE]);
var CODE_GENERATION_RESULT = {
  sources: new Map(),
  runtimeRequirements: new Set()
};

var CssModule =
/*#__PURE__*/
function (_webpack$default$Modu) {
  _inherits(CssModule, _webpack$default$Modu);

  function CssModule(_ref) {
    var _this;

    var context = _ref.context,
        identifier = _ref.identifier,
        identifierIndex = _ref.identifierIndex,
        content = _ref.content,
        media = _ref.media,
        sourceMap = _ref.sourceMap;

    _classCallCheck(this, CssModule);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CssModule).call(this, _utils.MODULE_TYPE, context));
    _this.id = '';
    _this._context = context;
    _this._identifier = identifier;
    _this._identifierIndex = identifierIndex;
    _this.content = content;
    _this.media = media;
    _this.sourceMap = sourceMap;
    _this.buildInfo = {};
    _this.buildMeta = {};
    return _this;
  } // no source() so webpack 4 doesn't do add stuff to the bundle


  _createClass(CssModule, [{
    key: "size",
    value: function size() {
      return this.content.length;
    }
  }, {
    key: "identifier",
    value: function identifier() {
      return "css ".concat(this._identifier, " ").concat(this._identifierIndex);
    }
  }, {
    key: "readableIdentifier",
    value: function readableIdentifier(requestShortener) {
      return "css ".concat(requestShortener.shorten(this._identifier)).concat(this._identifierIndex ? " (".concat(this._identifierIndex, ")") : '');
    } // eslint-disable-next-line class-methods-use-this

  }, {
    key: "getSourceTypes",
    value: function getSourceTypes() {
      return TYPES;
    } // eslint-disable-next-line class-methods-use-this

  }, {
    key: "codeGeneration",
    value: function codeGeneration() {
      return CODE_GENERATION_RESULT;
    }
  }, {
    key: "nameForCondition",
    value: function nameForCondition() {
      var resource = this._identifier.split('!').pop();

      var idx = resource.indexOf('?');

      if (idx >= 0) {
        return resource.substring(0, idx);
      }

      return resource;
    }
  }, {
    key: "updateCacheModule",
    value: function updateCacheModule(module) {
      this.content = module.content;
      this.media = module.media;
      this.sourceMap = module.sourceMap;
    } // eslint-disable-next-line class-methods-use-this

  }, {
    key: "needRebuild",
    value: function needRebuild() {
      return true;
    } // eslint-disable-next-line class-methods-use-this

  }, {
    key: "needBuild",
    value: function needBuild(context, callback) {
      callback(null, false);
    }
  }, {
    key: "build",
    value: function build(options, compilation, resolver, fileSystem, callback) {
      this.buildInfo = {};
      this.buildMeta = {};
      callback();
    }
  }, {
    key: "updateHash",
    value: function updateHash(hash, context) {
      _get(_getPrototypeOf(CssModule.prototype), "updateHash", this).call(this, hash, context);

      hash.update(this.content);
      hash.update(this.media || '');
      hash.update(this.sourceMap ? JSON.stringify(this.sourceMap) : '');
    }
  }, {
    key: "serialize",
    value: function serialize(context) {
      var write = context.write;
      write(this._context);
      write(this._identifier);
      write(this._identifierIndex);
      write(this.content);
      write(this.media);
      write(this.sourceMap);

      _get(_getPrototypeOf(CssModule.prototype), "serialize", this).call(this, context);
    }
  }, {
    key: "deserialize",
    value: function deserialize(context) {
      _get(_getPrototypeOf(CssModule.prototype), "deserialize", this).call(this, context);
    }
  }]);

  return CssModule;
}(_webpack["default"].Module);

if (_webpack["default"].util && _webpack["default"].util.serialization) {
  _webpack["default"].util.serialization.register(CssModule, 'mini-css-extract-plugin/dist/CssModule', null, {
    serialize: function serialize(instance, context) {
      instance.serialize(context);
    },
    deserialize: function deserialize(context) {
      var read = context.read;
      var dep = new CssModule({
        context: read(),
        identifier: read(),
        identifierIndex: read(),
        content: read(),
        media: read(),
        sourceMap: read()
      });
      dep.deserialize(context);
      return dep;
    }
  });
}

var _default = CssModule;
exports["default"] = _default;
