// ajv/dist/vocabularies/validation/dist/limitNumber.dev.js
"use strict";

function _templateObject3() {
  var data = _taggedTemplateLiteral(["", " ", " ", " || isNaN(", ")"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["{comparison: ", ", limit: ", "}"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["must be ", " ", ""]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

Object.defineProperty(exports, "__esModule", {
  value: true
});

var codegen_1 = require("../../compile/codegen");

var ops = codegen_1.operators;
var KWDs = {
  maximum: {
    okStr: "<=",
    ok: ops.LTE,
    fail: ops.GT
  },
  minimum: {
    okStr: ">=",
    ok: ops.GTE,
    fail: ops.LT
  },
  exclusiveMaximum: {
    okStr: "<",
    ok: ops.LT,
    fail: ops.GTE
  },
  exclusiveMinimum: {
    okStr: ">",
    ok: ops.GT,
    fail: ops.LTE
  }
};
var error = {
  message: function message(_ref) {
    var keyword = _ref.keyword,
        schemaCode = _ref.schemaCode;
    return (0, codegen_1.str)(_templateObject(), KWDs[keyword].okStr, schemaCode);
  },
  params: function params(_ref2) {
    var keyword = _ref2.keyword,
        schemaCode = _ref2.schemaCode;
    return (0, codegen_1._)(_templateObject2(), KWDs[keyword].okStr, schemaCode);
  }
};
var def = {
  keyword: Object.keys(KWDs),
  type: "number",
  schemaType: "number",
  $data: true,
  error: error,
  code: function code(cxt) {
    var keyword = cxt.keyword,
        data = cxt.data,
        schemaCode = cxt.schemaCode;
    cxt.fail$data((0, codegen_1._)(_templateObject3(), data, KWDs[keyword].fail, schemaCode, data));
  }
};
exports["default"] = def;
