// @eslint/eslintrc/lib/dist/index.dev.js
/**
 * @fileoverview Package exports for @eslint/eslintrc
 * @author Nicholas C. Zakas
 */
"use strict"; //------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

var _require = require("./config-array-factory"),
    ConfigArrayFactory = _require.ConfigArrayFactory,
    createConfigArrayFactoryContext = _require.createContext;

var _require2 = require("./cascading-config-array-factory"),
    CascadingConfigArrayFactory = _require2.CascadingConfigArrayFactory;

var ModuleResolver = require("./shared/relative-module-resolver");

var _require3 = require("./config-array"),
    ConfigArray = _require3.ConfigArray,
    getUsedExtractedConfigs = _require3.getUsedExtractedConfigs;

var _require4 = require("./config-array/config-dependency"),
    ConfigDependency = _require4.ConfigDependency;

var _require5 = require("./config-array/extracted-config"),
    ExtractedConfig = _require5.ExtractedConfig;

var _require6 = require("./config-array/ignore-pattern"),
    IgnorePattern = _require6.IgnorePattern;

var _require7 = require("./config-array/override-tester"),
    OverrideTester = _require7.OverrideTester;

var ConfigOps = require("./shared/config-ops");

var ConfigValidator = require("./shared/config-validator");

var naming = require("./shared/naming");

var _require8 = require("./flat-compat"),
    FlatCompat = _require8.FlatCompat; //-----------------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------------


module.exports = {
  Legacy: {
    ConfigArray: ConfigArray,
    createConfigArrayFactoryContext: createConfigArrayFactoryContext,
    CascadingConfigArrayFactory: CascadingConfigArrayFactory,
    ConfigArrayFactory: ConfigArrayFactory,
    ConfigDependency: ConfigDependency,
    ExtractedConfig: ExtractedConfig,
    IgnorePattern: IgnorePattern,
    OverrideTester: OverrideTester,
    getUsedExtractedConfigs: getUsedExtractedConfigs,
    // shared
    ConfigOps: ConfigOps,
    ConfigValidator: ConfigValidator,
    ModuleResolver: ModuleResolver,
    naming: naming
  },
  FlatCompat: FlatCompat
};
