// react-router/modules/dist/index.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MemoryRouter", {
  enumerable: true,
  get: function get() {
    return _MemoryRouter["default"];
  }
});
Object.defineProperty(exports, "Prompt", {
  enumerable: true,
  get: function get() {
    return _Prompt["default"];
  }
});
Object.defineProperty(exports, "Redirect", {
  enumerable: true,
  get: function get() {
    return _Redirect["default"];
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function get() {
    return _Route["default"];
  }
});
Object.defineProperty(exports, "Router", {
  enumerable: true,
  get: function get() {
    return _Router["default"];
  }
});
Object.defineProperty(exports, "StaticRouter", {
  enumerable: true,
  get: function get() {
    return _StaticRouter["default"];
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function get() {
    return _Switch["default"];
  }
});
Object.defineProperty(exports, "generatePath", {
  enumerable: true,
  get: function get() {
    return _generatePath["default"];
  }
});
Object.defineProperty(exports, "matchPath", {
  enumerable: true,
  get: function get() {
    return _matchPath["default"];
  }
});
Object.defineProperty(exports, "withRouter", {
  enumerable: true,
  get: function get() {
    return _withRouter["default"];
  }
});
Object.defineProperty(exports, "useHistory", {
  enumerable: true,
  get: function get() {
    return _hooks.useHistory;
  }
});
Object.defineProperty(exports, "useLocation", {
  enumerable: true,
  get: function get() {
    return _hooks.useLocation;
  }
});
Object.defineProperty(exports, "useParams", {
  enumerable: true,
  get: function get() {
    return _hooks.useParams;
  }
});
Object.defineProperty(exports, "useRouteMatch", {
  enumerable: true,
  get: function get() {
    return _hooks.useRouteMatch;
  }
});
Object.defineProperty(exports, "__HistoryContext", {
  enumerable: true,
  get: function get() {
    return _HistoryContext["default"];
  }
});
Object.defineProperty(exports, "__RouterContext", {
  enumerable: true,
  get: function get() {
    return _RouterContext["default"];
  }
});

var _MemoryRouter = _interopRequireDefault(require("./MemoryRouter.js"));

var _Prompt = _interopRequireDefault(require("./Prompt.js"));

var _Redirect = _interopRequireDefault(require("./Redirect.js"));

var _Route = _interopRequireDefault(require("./Route.js"));

var _Router = _interopRequireDefault(require("./Router.js"));

var _StaticRouter = _interopRequireDefault(require("./StaticRouter.js"));

var _Switch = _interopRequireDefault(require("./Switch.js"));

var _generatePath = _interopRequireDefault(require("./generatePath.js"));

var _matchPath = _interopRequireDefault(require("./matchPath.js"));

var _withRouter = _interopRequireDefault(require("./withRouter.js"));

var _hooks = require("./hooks.js");

var _HistoryContext = _interopRequireDefault(require("./HistoryContext.js"));

var _RouterContext = _interopRequireDefault(require("./RouterContext.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

if (__DEV__) {
  if (typeof window !== "undefined") {
    var global = window;
    var key = "__react_router_build__";
    var buildNames = {
      cjs: "CommonJS",
      esm: "ES modules",
      umd: "UMD"
    };

    if (global[key] && global[key] !== process.env.BUILD_FORMAT) {
      var initialBuildName = buildNames[global[key]];
      var secondaryBuildName = buildNames[process.env.BUILD_FORMAT]; // TODO: Add link to article that explains in detail how to avoid
      // loading 2 different builds.

      throw new Error("You are loading the ".concat(secondaryBuildName, " build of React Router ") + "on a page that is already running the ".concat(initialBuildName, " ") + "build, so things won't work right.");
    }

    global[key] = process.env.BUILD_FORMAT;
  }
}
