// workbox-strategies/src/dist/CacheFirst.js
"use strict";
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.CacheFirst = void 0;
var assert_js_1 = require("workbox-core/_private/assert.js");
var cacheNames_js_1 = require("workbox-core/_private/cacheNames.js");
var cacheWrapper_js_1 = require("workbox-core/_private/cacheWrapper.js");
var fetchWrapper_js_1 = require("workbox-core/_private/fetchWrapper.js");
var getFriendlyURL_js_1 = require("workbox-core/_private/getFriendlyURL.js");
var logger_js_1 = require("workbox-core/_private/logger.js");
var WorkboxError_js_1 = require("workbox-core/_private/WorkboxError.js");
var messages_js_1 = require("./utils/messages.js");
require("./_version.js");
/**
 * An implementation of a [cache-first]{@link https://developers.google.com/web/fundamentals/instant-and-offline/offline-cookbook/#cache-falling-back-to-network}
 * request strategy.
 *
 * A cache first strategy is useful for assets that have been revisioned,
 * such as URLs like `/styles/example.a8f5f1.css`, since they
 * can be cached for long periods of time.
 *
 * If the network request fails, and there is no cache match, this will throw
 * a `WorkboxError` exception.
 *
 * @memberof module:workbox-strategies
 */
var CacheFirst = /** @class */ (function () {
    /**
     * @param {Object} options
     * @param {string} options.cacheName Cache name to store and retrieve
     * requests. Defaults to cache names provided by
     * [workbox-core]{@link module:workbox-core.cacheNames}.
     * @param {Array<Object>} options.plugins [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins}
     * to use in conjunction with this caching strategy.
     * @param {Object} options.fetchOptions Values passed along to the
     * [`init`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters)
     * of all fetch() requests made by this strategy.
     * @param {Object} options.matchOptions [`CacheQueryOptions`](https://w3c.github.io/ServiceWorker/#dictdef-cachequeryoptions)
     */
    function CacheFirst(options) {
        if (options === void 0) { options = {}; }
        this._cacheName = cacheNames_js_1.cacheNames.getRuntimeName(options.cacheName);
        this._plugins = options.plugins || [];
        this._fetchOptions = options.fetchOptions;
        this._matchOptions = options.matchOptions;
    }
    /**
     * This method will perform a request strategy and follows an API that
     * will work with the
     * [Workbox Router]{@link module:workbox-routing.Router}.
     *
     * @param {Object} options
     * @param {Request|string} options.request A request to run this strategy for.
     * @param {Event} [options.event] The event that triggered the request.
     * @return {Promise<Response>}
     */
    CacheFirst.prototype.handle = function (_a) {
        var event = _a.event, request = _a.request;
        return __awaiter(this, void 0, Promise, function () {
            var logs, response, error, err_1, _i, logs_1, log;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        logs = [];
                        if (typeof request === 'string') {
                            request = new Request(request);
                        }
                        if (process.env.NODE_ENV !== 'production') {
                            assert_js_1.assert.isInstance(request, Request, {
                                moduleName: 'workbox-strategies',
                                className: 'CacheFirst',
                                funcName: 'makeRequest',
                                paramName: 'request'
                            });
                        }
                        return [4 /*yield*/, cacheWrapper_js_1.cacheWrapper.match({
                                cacheName: this._cacheName,
                                request: request,
                                event: event,
                                matchOptions: this._matchOptions,
                                plugins: this._plugins
                            })];
                    case 1:
                        response = _b.sent();
                        if (!!response) return [3 /*break*/, 6];
                        if (process.env.NODE_ENV !== 'production') {
                            logs.push("No response found in the '" + this._cacheName + "' cache. " +
                                "Will respond with a network request.");
                        }
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 4, , 5]);
                        return [4 /*yield*/, this._getFromNetwork(request, event)];
                    case 3:
                        response = _b.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        err_1 = _b.sent();
                        error = err_1;
                        return [3 /*break*/, 5];
                    case 5:
                        if (process.env.NODE_ENV !== 'production') {
                            if (response) {
                                logs.push("Got response from network.");
                            }
                            else {
                                logs.push("Unable to get a response from the network.");
                            }
                        }
                        return [3 /*break*/, 7];
                    case 6:
                        if (process.env.NODE_ENV !== 'production') {
                            logs.push("Found a cached response in the '" + this._cacheName + "' cache.");
                        }
                        _b.label = 7;
                    case 7:
                        if (process.env.NODE_ENV !== 'production') {
                            logger_js_1.logger.groupCollapsed(messages_js_1.messages.strategyStart('CacheFirst', request));
                            for (_i = 0, logs_1 = logs; _i < logs_1.length; _i++) {
                                log = logs_1[_i];
                                logger_js_1.logger.log(log);
                            }
                            messages_js_1.messages.printFinalResponse(response);
                            logger_js_1.logger.groupEnd();
                        }
                        if (!response) {
                            throw new WorkboxError_js_1.WorkboxError('no-response', { url: request.url, error: error });
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * Handles the network and cache part of CacheFirst.
     *
     * @param {Request} request
     * @param {Event} [event]
     * @return {Promise<Response>}
     *
     * @private
     */
    CacheFirst.prototype._getFromNetwork = function (request, event) {
        return __awaiter(this, void 0, void 0, function () {
            var response, responseClone, cachePutPromise;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, fetchWrapper_js_1.fetchWrapper.fetch({
                            request: request,
                            event: event,
                            fetchOptions: this._fetchOptions,
                            plugins: this._plugins
                        })];
                    case 1:
                        response = _a.sent();
                        responseClone = response.clone();
                        cachePutPromise = cacheWrapper_js_1.cacheWrapper.put({
                            cacheName: this._cacheName,
                            request: request,
                            response: responseClone,
                            event: event,
                            plugins: this._plugins
                        });
                        if (event) {
                            try {
                                event.waitUntil(cachePutPromise);
                            }
                            catch (error) {
                                if (process.env.NODE_ENV !== 'production') {
                                    logger_js_1.logger.warn("Unable to ensure service worker stays alive when " +
                                        ("updating cache for '" + getFriendlyURL_js_1.getFriendlyURL(request.url) + "'."));
                                }
                            }
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    return CacheFirst;
}());
exports.CacheFirst = CacheFirst;
