// workbox-strategies/src/dist/StaleWhileRevalidate.js
"use strict";
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
exports.__esModule = true;
exports.StaleWhileRevalidate = void 0;
var assert_js_1 = require("workbox-core/_private/assert.js");
var cacheNames_js_1 = require("workbox-core/_private/cacheNames.js");
var cacheWrapper_js_1 = require("workbox-core/_private/cacheWrapper.js");
var fetchWrapper_js_1 = require("workbox-core/_private/fetchWrapper.js");
var getFriendlyURL_js_1 = require("workbox-core/_private/getFriendlyURL.js");
var logger_js_1 = require("workbox-core/_private/logger.js");
var WorkboxError_js_1 = require("workbox-core/_private/WorkboxError.js");
var messages_js_1 = require("./utils/messages.js");
var cacheOkAndOpaquePlugin_js_1 = require("./plugins/cacheOkAndOpaquePlugin.js");
require("./_version.js");
/**
 * An implementation of a
 * [stale-while-revalidate]{@link https://developers.google.com/web/fundamentals/instant-and-offline/offline-cookbook/#stale-while-revalidate}
 * request strategy.
 *
 * Resources are requested from both the cache and the network in parallel.
 * The strategy will respond with the cached version if available, otherwise
 * wait for the network response. The cache is updated with the network response
 * with each successful request.
 *
 * By default, this strategy will cache responses with a 200 status code as
 * well as [opaque responses]{@link https://developers.google.com/web/tools/workbox/guides/handle-third-party-requests}.
 * Opaque responses are cross-origin requests where the response doesn't
 * support [CORS]{@link https://enable-cors.org/}.
 *
 * If the network request fails, and there is no cache match, this will throw
 * a `WorkboxError` exception.
 *
 * @memberof module:workbox-strategies
 */
var StaleWhileRevalidate = /** @class */ (function () {
    /**
     * @param {Object} options
     * @param {string} options.cacheName Cache name to store and retrieve
     * requests. Defaults to cache names provided by
     * [workbox-core]{@link module:workbox-core.cacheNames}.
     * @param {Array<Object>} options.plugins [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins}
     * to use in conjunction with this caching strategy.
     * @param {Object} options.fetchOptions Values passed along to the
     * [`init`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters)
     * of all fetch() requests made by this strategy.
     * @param {Object} options.matchOptions [`CacheQueryOptions`](https://w3c.github.io/ServiceWorker/#dictdef-cachequeryoptions)
     */
    function StaleWhileRevalidate(options) {
        if (options === void 0) { options = {}; }
        this._cacheName = cacheNames_js_1.cacheNames.getRuntimeName(options.cacheName);
        this._plugins = options.plugins || [];
        if (options.plugins) {
            var isUsingCacheWillUpdate = options.plugins.some(function (plugin) { return !!plugin.cacheWillUpdate; });
            this._plugins = isUsingCacheWillUpdate ?
                options.plugins : __spreadArrays([cacheOkAndOpaquePlugin_js_1.cacheOkAndOpaquePlugin], options.plugins);
        }
        else {
            // No plugins passed in, use the default plugin.
            this._plugins = [cacheOkAndOpaquePlugin_js_1.cacheOkAndOpaquePlugin];
        }
        this._fetchOptions = options.fetchOptions;
        this._matchOptions = options.matchOptions;
    }
    /**
     * This method will perform a request strategy and follows an API that
     * will work with the
     * [Workbox Router]{@link module:workbox-routing.Router}.
     *
     * @param {Object} options
     * @param {Request|string} options.request A request to run this strategy for.
     * @param {Event} [options.event] The event that triggered the request.
     * @return {Promise<Response>}
     */
    StaleWhileRevalidate.prototype.handle = function (_a) {
        var event = _a.event, request = _a.request;
        return __awaiter(this, void 0, Promise, function () {
            var logs, fetchAndCachePromise, response, error, err_1, _i, logs_1, log;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        logs = [];
                        if (typeof request === 'string') {
                            request = new Request(request);
                        }
                        if (process.env.NODE_ENV !== 'production') {
                            assert_js_1.assert.isInstance(request, Request, {
                                moduleName: 'workbox-strategies',
                                className: 'StaleWhileRevalidate',
                                funcName: 'handle',
                                paramName: 'request'
                            });
                        }
                        fetchAndCachePromise = this._getFromNetwork({ request: request, event: event });
                        return [4 /*yield*/, cacheWrapper_js_1.cacheWrapper.match({
                                cacheName: this._cacheName,
                                request: request,
                                event: event,
                                matchOptions: this._matchOptions,
                                plugins: this._plugins
                            })];
                    case 1:
                        response = _b.sent();
                        if (!response) return [3 /*break*/, 2];
                        if (process.env.NODE_ENV !== 'production') {
                            logs.push("Found a cached response in the '" + this._cacheName + "'" +
                                " cache. Will update with the network response in the background.");
                        }
                        if (event) {
                            try {
                                event.waitUntil(fetchAndCachePromise);
                            }
                            catch (error) {
                                if (process.env.NODE_ENV !== 'production') {
                                    logger_js_1.logger.warn("Unable to ensure service worker stays alive when " +
                                        ("updating cache for '" + getFriendlyURL_js_1.getFriendlyURL(request.url) + "'."));
                                }
                            }
                        }
                        return [3 /*break*/, 6];
                    case 2:
                        if (process.env.NODE_ENV !== 'production') {
                            logs.push("No response found in the '" + this._cacheName + "' cache. " +
                                "Will wait for the network response.");
                        }
                        _b.label = 3;
                    case 3:
                        _b.trys.push([3, 5, , 6]);
                        return [4 /*yield*/, fetchAndCachePromise];
                    case 4:
                        response = _b.sent();
                        return [3 /*break*/, 6];
                    case 5:
                        err_1 = _b.sent();
                        error = err_1;
                        return [3 /*break*/, 6];
                    case 6:
                        if (process.env.NODE_ENV !== 'production') {
                            logger_js_1.logger.groupCollapsed(messages_js_1.messages.strategyStart('StaleWhileRevalidate', request));
                            for (_i = 0, logs_1 = logs; _i < logs_1.length; _i++) {
                                log = logs_1[_i];
                                logger_js_1.logger.log(log);
                            }
                            messages_js_1.messages.printFinalResponse(response);
                            logger_js_1.logger.groupEnd();
                        }
                        if (!response) {
                            throw new WorkboxError_js_1.WorkboxError('no-response', { url: request.url, error: error });
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * @param {Object} options
     * @param {Request} options.request
     * @param {Event} [options.event]
     * @return {Promise<Response>}
     *
     * @private
     */
    StaleWhileRevalidate.prototype._getFromNetwork = function (_a) {
        var request = _a.request, event = _a.event;
        return __awaiter(this, void 0, Promise, function () {
            var response, cachePutPromise;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, fetchWrapper_js_1.fetchWrapper.fetch({
                            request: request,
                            event: event,
                            fetchOptions: this._fetchOptions,
                            plugins: this._plugins
                        })];
                    case 1:
                        response = _b.sent();
                        cachePutPromise = cacheWrapper_js_1.cacheWrapper.put({
                            cacheName: this._cacheName,
                            request: request,
                            response: response.clone(),
                            event: event,
                            plugins: this._plugins
                        });
                        if (event) {
                            try {
                                event.waitUntil(cachePutPromise);
                            }
                            catch (error) {
                                if (process.env.NODE_ENV !== 'production') {
                                    logger_js_1.logger.warn("Unable to ensure service worker stays alive when " +
                                        ("updating cache for '" + getFriendlyURL_js_1.getFriendlyURL(request.url) + "'."));
                                }
                            }
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    return StaleWhileRevalidate;
}());
exports.StaleWhileRevalidate = StaleWhileRevalidate;
