// @babel/preset-env/lib/dist/index.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isPluginRequired = isPluginRequired;
exports["default"] = exports.getPolyfillPlugins = exports.getModulesPluginNames = exports.transformIncludesAndExcludes = void 0;

var _semver = require("semver");

var _debug = require("./debug");

var _getOptionSpecificExcludes = require("./get-option-specific-excludes");

var _filterItems = require("./filter-items");

var _moduleTransformations = require("./module-transformations");

var _normalizeOptions = require("./normalize-options");

var _shippedProposals = require("../data/shipped-proposals");

var _pluginsCompatData = require("./plugins-compat-data");

var _overlappingPlugins = require("@babel/compat-data/overlapping-plugins");

var _regenerator = require("./polyfills/regenerator");

var _babelPolyfill = require("./polyfills/babel-polyfill");

var _babelPluginPolyfillCorejs = require("babel-plugin-polyfill-corejs2");

var _babelPluginPolyfillCorejs2 = require("babel-plugin-polyfill-corejs3");

var _babelPluginPolyfillRegenerator = require("babel-plugin-polyfill-regenerator");

var _helperCompilationTargets = require("@babel/helper-compilation-targets");

var _availablePlugins = require("./available-plugins");

var _helperPluginUtils = require("@babel/helper-plugin-utils");

var pluginCoreJS2 = _babelPluginPolyfillCorejs["default"] || _babelPluginPolyfillCorejs;
var pluginCoreJS3 = _babelPluginPolyfillCorejs2["default"] || _babelPluginPolyfillCorejs2;
var pluginRegenerator = _babelPluginPolyfillRegenerator["default"] || _babelPluginPolyfillRegenerator;

function isPluginRequired(targets, support) {
  return (0, _helperCompilationTargets.isRequired)("fake-name", targets, {
    compatData: {
      "fake-name": support
    }
  });
}

function filterStageFromList(list, stageList) {
  return Object.keys(list).reduce(function (result, item) {
    if (!stageList.has(item)) {
      result[item] = list[item];
    }

    return result;
  }, {});
}

var pluginLists = {
  withProposals: {
    withoutBugfixes: _pluginsCompatData.plugins,
    withBugfixes: Object.assign({}, _pluginsCompatData.plugins, _pluginsCompatData.pluginsBugfixes)
  },
  withoutProposals: {
    withoutBugfixes: filterStageFromList(_pluginsCompatData.plugins, _shippedProposals.proposalPlugins),
    withBugfixes: filterStageFromList(Object.assign({}, _pluginsCompatData.plugins, _pluginsCompatData.pluginsBugfixes), _shippedProposals.proposalPlugins)
  }
};

function getPluginList(proposals, bugfixes) {
  if (proposals) {
    if (bugfixes) return pluginLists.withProposals.withBugfixes;else return pluginLists.withProposals.withoutBugfixes;
  } else {
    if (bugfixes) return pluginLists.withoutProposals.withBugfixes;else return pluginLists.withoutProposals.withoutBugfixes;
  }
}

var getPlugin = function getPlugin(pluginName) {
  var plugin = _availablePlugins["default"][pluginName]();

  if (!plugin) {
    throw new Error("Could not find plugin \"".concat(pluginName, "\". Ensure there is an entry in ./available-plugins.js for it."));
  }

  return plugin;
};

var transformIncludesAndExcludes = function transformIncludesAndExcludes(opts) {
  return opts.reduce(function (result, opt) {
    var target = opt.match(/^(es|es6|es7|esnext|web)\./) ? "builtIns" : "plugins";
    result[target].add(opt);
    return result;
  }, {
    all: opts,
    plugins: new Set(),
    builtIns: new Set()
  });
};

exports.transformIncludesAndExcludes = transformIncludesAndExcludes;

var getModulesPluginNames = function getModulesPluginNames(_ref) {
  var modules = _ref.modules,
      transformations = _ref.transformations,
      shouldTransformESM = _ref.shouldTransformESM,
      shouldTransformDynamicImport = _ref.shouldTransformDynamicImport,
      shouldTransformExportNamespaceFrom = _ref.shouldTransformExportNamespaceFrom,
      shouldParseTopLevelAwait = _ref.shouldParseTopLevelAwait;
  var modulesPluginNames = [];

  if (modules !== false && transformations[modules]) {
    if (shouldTransformESM) {
      modulesPluginNames.push(transformations[modules]);
    }

    if (shouldTransformDynamicImport && shouldTransformESM && modules !== "umd") {
      modulesPluginNames.push("proposal-dynamic-import");
    } else {
      if (shouldTransformDynamicImport) {
        console.warn("Dynamic import can only be supported when transforming ES modules" + " to AMD, CommonJS or SystemJS. Only the parser plugin will be enabled.");
      }

      modulesPluginNames.push("syntax-dynamic-import");
    }
  } else {
    modulesPluginNames.push("syntax-dynamic-import");
  }

  if (shouldTransformExportNamespaceFrom) {
    modulesPluginNames.push("proposal-export-namespace-from");
  } else {
    modulesPluginNames.push("syntax-export-namespace-from");
  }

  if (shouldParseTopLevelAwait) {
    modulesPluginNames.push("syntax-top-level-await");
  }

  return modulesPluginNames;
};

exports.getModulesPluginNames = getModulesPluginNames;

var getPolyfillPlugins = function getPolyfillPlugins(_ref2) {
  var useBuiltIns = _ref2.useBuiltIns,
      corejs = _ref2.corejs,
      polyfillTargets = _ref2.polyfillTargets,
      include = _ref2.include,
      exclude = _ref2.exclude,
      proposals = _ref2.proposals,
      shippedProposals = _ref2.shippedProposals,
      regenerator = _ref2.regenerator,
      debug = _ref2.debug;
  var polyfillPlugins = [];

  if (useBuiltIns === "usage" || useBuiltIns === "entry") {
    var pluginOptions = {
      method: "".concat(useBuiltIns, "-global"),
      version: corejs ? corejs.toString() : undefined,
      targets: polyfillTargets,
      include: include,
      exclude: exclude,
      proposals: proposals,
      shippedProposals: shippedProposals,
      debug: debug
    };

    if (corejs) {
      if (useBuiltIns === "usage") {
        if (corejs.major === 2) {
          polyfillPlugins.push([pluginCoreJS2, pluginOptions], [_babelPolyfill["default"], {
            usage: true
          }]);
        } else {
          polyfillPlugins.push([pluginCoreJS3, pluginOptions], [_babelPolyfill["default"], {
            usage: true,
            deprecated: true
          }]);
        }

        if (regenerator) {
          polyfillPlugins.push([pluginRegenerator, {
            method: "usage-global",
            debug: debug
          }]);
        }
      } else {
        if (corejs.major === 2) {
          polyfillPlugins.push([_babelPolyfill["default"], {
            regenerator: regenerator
          }], [pluginCoreJS2, pluginOptions]);
        } else {
          polyfillPlugins.push([pluginCoreJS3, pluginOptions], [_babelPolyfill["default"], {
            deprecated: true
          }]);

          if (!regenerator) {
            polyfillPlugins.push([_regenerator["default"], pluginOptions]);
          }
        }
      }
    }
  }

  return polyfillPlugins;
};

exports.getPolyfillPlugins = getPolyfillPlugins;

function getLocalTargets(optionsTargets, ignoreBrowserslistConfig, configPath, browserslistEnv) {
  if (optionsTargets != null && optionsTargets.esmodules && optionsTargets.browsers) {
    console.warn("\n@babel/preset-env: esmodules and browsers targets have been specified together.\n`browsers` target, `".concat(optionsTargets.browsers.toString(), "` will be ignored.\n"));
  }

  return (0, _helperCompilationTargets["default"])(optionsTargets, {
    ignoreBrowserslistConfig: ignoreBrowserslistConfig,
    configPath: configPath,
    browserslistEnv: browserslistEnv
  });
}

function supportsStaticESM(caller) {
  return !!(caller != null && caller.supportsStaticESM);
}

function supportsDynamicImport(caller) {
  return !!(caller != null && caller.supportsDynamicImport);
}

function supportsExportNamespaceFrom(caller) {
  return !!(caller != null && caller.supportsExportNamespaceFrom);
}

function supportsTopLevelAwait(caller) {
  return !!(caller != null && caller.supportsTopLevelAwait);
}

var _default = (0, _helperPluginUtils.declare)(function (api, opts) {
  api.assertVersion(7);
  var babelTargets = api.targets();

  var _ref3 = (0, _normalizeOptions["default"])(opts),
      bugfixes = _ref3.bugfixes,
      configPath = _ref3.configPath,
      debug = _ref3.debug,
      optionsExclude = _ref3.exclude,
      forceAllTransforms = _ref3.forceAllTransforms,
      ignoreBrowserslistConfig = _ref3.ignoreBrowserslistConfig,
      optionsInclude = _ref3.include,
      loose = _ref3.loose,
      modules = _ref3.modules,
      shippedProposals = _ref3.shippedProposals,
      spec = _ref3.spec,
      optionsTargets = _ref3.targets,
      useBuiltIns = _ref3.useBuiltIns,
      _ref3$corejs = _ref3.corejs,
      corejs = _ref3$corejs.version,
      proposals = _ref3$corejs.proposals,
      browserslistEnv = _ref3.browserslistEnv;

  var targets = babelTargets;

  if ((0, _semver.lt)(api.version, "7.13.0") || opts.targets || opts.configPath || opts.browserslistEnv || opts.ignoreBrowserslistConfig) {
    {
      var hasUglifyTarget = false;

      if (optionsTargets != null && optionsTargets.uglify) {
        hasUglifyTarget = true;
        delete optionsTargets.uglify;
        console.warn("\nThe uglify target has been deprecated. Set the top level\noption `forceAllTransforms: true` instead.\n");
      }
    }
    targets = getLocalTargets(optionsTargets, ignoreBrowserslistConfig, configPath, browserslistEnv);
  }

  var transformTargets = forceAllTransforms || hasUglifyTarget ? {} : targets;
  var include = transformIncludesAndExcludes(optionsInclude);
  var exclude = transformIncludesAndExcludes(optionsExclude);
  var compatData = getPluginList(shippedProposals, bugfixes);
  var shouldSkipExportNamespaceFrom = modules === "auto" && (api.caller == null ? void 0 : api.caller(supportsExportNamespaceFrom)) || modules === false && !(0, _helperCompilationTargets.isRequired)("proposal-export-namespace-from", transformTargets, {
    compatData: compatData,
    includes: include.plugins,
    excludes: exclude.plugins
  });
  var modulesPluginNames = getModulesPluginNames({
    modules: modules,
    transformations: _moduleTransformations["default"],
    shouldTransformESM: modules !== "auto" || !(api.caller != null && api.caller(supportsStaticESM)),
    shouldTransformDynamicImport: modules !== "auto" || !(api.caller != null && api.caller(supportsDynamicImport)),
    shouldTransformExportNamespaceFrom: !shouldSkipExportNamespaceFrom,
    shouldParseTopLevelAwait: !api.caller || api.caller(supportsTopLevelAwait)
  });
  var pluginNames = (0, _helperCompilationTargets.filterItems)(compatData, include.plugins, exclude.plugins, transformTargets, modulesPluginNames, (0, _getOptionSpecificExcludes["default"])({
    loose: loose
  }), _shippedProposals.pluginSyntaxMap);
  (0, _filterItems.removeUnnecessaryItems)(pluginNames, _overlappingPlugins);
  (0, _filterItems.removeUnsupportedItems)(pluginNames, api.version);
  var polyfillPlugins = getPolyfillPlugins({
    useBuiltIns: useBuiltIns,
    corejs: corejs,
    polyfillTargets: targets,
    include: include.builtIns,
    exclude: exclude.builtIns,
    proposals: proposals,
    shippedProposals: shippedProposals,
    regenerator: pluginNames.has("transform-regenerator"),
    debug: debug
  });
  var pluginUseBuiltIns = useBuiltIns !== false;
  var plugins = Array.from(pluginNames).map(function (pluginName) {
    if (pluginName === "proposal-class-properties" || pluginName === "proposal-private-methods" || pluginName === "proposal-private-property-in-object") {
      return [getPlugin(pluginName), {
        loose: loose ? "#__internal__@babel/preset-env__prefer-true-but-false-is-ok-if-it-prevents-an-error" : "#__internal__@babel/preset-env__prefer-false-but-true-is-ok-if-it-prevents-an-error"
      }];
    }

    return [getPlugin(pluginName), {
      spec: spec,
      loose: loose,
      useBuiltIns: pluginUseBuiltIns
    }];
  }).concat(polyfillPlugins);

  if (debug) {
    console.log("@babel/preset-env: `DEBUG` option");
    console.log("\nUsing targets:");
    console.log(JSON.stringify((0, _helperCompilationTargets.prettifyTargets)(targets), null, 2));
    console.log("\nUsing modules transform: ".concat(modules.toString()));
    console.log("\nUsing plugins:");
    pluginNames.forEach(function (pluginName) {
      (0, _debug.logPlugin)(pluginName, targets, compatData);
    });

    if (!useBuiltIns) {
      console.log("\nUsing polyfills: No polyfills were added, since the `useBuiltIns` option was not set.");
    }
  }

  return {
    plugins: plugins
  };
});

exports["default"] = _default;
