// @babel/helper-define-polyfill-provider/lib/node/dist/dependencies.dev.js
"use strict";

exports.__esModule = true;
exports.resolve = resolve;
exports.has = has;
exports.logMissing = logMissing;
exports.laterLogMissing = laterLogMissing;

var _path = _interopRequireDefault(require("path"));

var _lodash = _interopRequireDefault(require("lodash.debounce"));

var _resolve = _interopRequireDefault(require("resolve"));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

var nativeRequireResolve = parseFloat(process.versions.node) >= 8.9;

function resolve(dirname, moduleName, absoluteImports) {
  if (absoluteImports === false) return moduleName;
  var basedir = dirname;

  if (typeof absoluteImports === "string") {
    basedir = _path["default"].resolve(basedir, absoluteImports);
  }

  var modulePackage, moduleNestedPath;
  var slash = moduleName.indexOf("/");

  if (moduleName[0] === "@") {
    slash = moduleName.indexOf("/", slash + 1);
  }

  if (slash === -1) {
    modulePackage = moduleName;
    moduleNestedPath = "";
  } else {
    modulePackage = moduleName.slice(0, slash);
    moduleNestedPath = moduleName.slice(slash);
  }

  try {
    var pkg;

    if (nativeRequireResolve) {
      // $FlowIgnore
      pkg = require.resolve("".concat(modulePackage, "/package.json"), {
        paths: [basedir]
      });
    } else {
      pkg = _resolve["default"].sync("".concat(modulePackage, "/package.json"), {
        basedir: basedir
      });
    }

    return _path["default"].dirname(pkg) + moduleNestedPath;
  } catch (err) {
    if (err.code !== "MODULE_NOT_FOUND") throw err; // $FlowIgnore

    throw Object.assign(new Error("Failed to resolve \"".concat(moduleName, "\" relative to \"").concat(dirname, "\"")), {
      code: "BABEL_POLYFILL_NOT_FOUND",
      polyfill: moduleName,
      dirname: dirname
    });
  }
}

function has(basedir, name) {
  try {
    if (nativeRequireResolve) {
      // $FlowIgnore
      require.resolve(name, {
        paths: [basedir]
      });
    } else {
      _resolve["default"].sync(name, {
        basedir: basedir
      });
    }

    return true;
  } catch (_unused) {
    return false;
  }
}

function logMissing(missingDeps) {
  if (missingDeps.size === 0) return;
  var deps = Array.from(missingDeps).sort().join(" ");
  console.warn("\nSome polyfills have been added but are not present in your dependencies.\n" + "Please run one of the following commands:\n" + "\tnpm install --save ".concat(deps, "\n") + "\tyarn add ".concat(deps, "\n"));
  process.exitCode = 1;
}

var allMissingDeps = new Set();
var laterLogMissingDependencies = (0, _lodash["default"])(function () {
  logMissing(allMissingDeps);
  allMissingDeps = new Set();
}, 100);

function laterLogMissing(missingDeps) {
  if (missingDeps.size === 0) return;
  missingDeps.forEach(function (name) {
    return allMissingDeps.add(name);
  });
  laterLogMissingDependencies();
}
